### ISSUE TITLE
Push up to commit

### ISSUE THREAD

**Issue by pank** [magit/magit#2084]
Hi,

I hope this is not a duplicate.

I would like to push up to commit `SOMETHING` (some sha1).  The corresponding git command would be 

```
git push origin SOMETHING:branch
```

I use this when I want to push some patches earlier than others.

Rasmus

**Comment by Silex** [comment]
When I want to do that, I go on the commit, the I press `P o` (push other), followed by enter because the completion is correct, and then I select the appropriate branch name.

**Comment by pank** [comment]
> When I want to do that, I go on the commit, the I press `P o` (push
> other), followed by enter because the completion is correct, and then
> I select the appropriate branch name.

I cannot reproduce.

I do the following.

   1. Go to my favorite unpushed commit.
   2. From the file view I click `P -d o`.
      - I'm asked about "Push".  No idea what Push means here, but
        Completion suggests a branch.
      - When I click `RET` I'm told "Nothing selected".  So I
        enter `master`
   3. `$` suggests the following, which is not a limited range, but all unpushed commits.

```
       0 git … push -v --dry-run origin master:refs/heads/master
     Pushing to orgmode@orgmode.org:org-mode.git
     To orgmode@orgmode.org:org-mode.git
        ed6349d..f2d1cba  master -> master
```


I simply want to put point at the last commit in my list of unpushed
commits that I want to push ATM and click a button.

**Comment by Silex** [comment]
> I'm asked about "Push". No idea what Push means here, but Completion suggests a branch.

By me it offers the SHA1 of the commit the point is on.

**Comment by kyleam** [comment]
Yes, it doesn't seem to work with standard completion.  This patch adds
the commit at point as the default if others fail.

```diff
diff --git a/lisp/magit-remote.el b/lisp/magit-remote.el
index e22222a..ecc936f 100644
--- a/lisp/magit-remote.el
+++ b/lisp/magit-remote.el
@@ -213,7 +213,8 @@ (defun magit-push-read-args (&optional use-upstream use-current default-current)
                               (magit-list-local-branch-names))
                      nil nil nil 'magit-revision-history
                      (or (and default-current current)
-                         (magit-local-branch-at-point)))
+                         (magit-local-branch-at-point)
+                         (magit-commit-at-point)))
                     (user-error "Nothing selected")))
          (remote (and (magit-branch-p local)
                       (magit-get-remote-branch local))))
```

**Comment by tarsius** [comment]
@kyleam In that case, please apply this change.

---

Mentioning #1485, because the point raised here is yet another point I have to remember when tackling that.

**Comment by tarsius** [comment]
I have installed this change. This will have to do for now.

**Comment by Silex** [comment]
Did it work by me because I'm using ido's completion?

**Comment by kyleam** [comment]
> Did it work by me because I'm using ido's completion?

Yes.  It worked for me too under helm, but not with 'emacs -Q'.

**Comment by tarsius** [comment]
> It worked for me too under helm,

So you use `helm-completing-read-with-cands-in-buffer` as `magit-completing-read-function` and that never causes any problems? If so could you please find the places where we claim that only Ido is supported and adjust it accordingly?

**Comment by pank** [comment]
Philippe Vaucher <notifications@github.com> writes:

> Did it work by me because I'm using ido's completion?

I use ido (`ido-everywhere => t`), though I have put no effort in telling
magit explicit that ido is my preferred completion mechanism.  I does not
seem to have picked it up, though.

Rasmus

-- 
⠠⠵

**Comment by tarsius** [comment]
There's an option for that (tm) :-)

**Comment by pank** [comment]
Jonas Bernoulli <notifications@github.com> writes:

> There's an option for that (tm) :-)

OK, thanks, I didn't know.

—Rasmus, who also wants to kill `org-completion-use-ido' ASAP.

-- 
However beautiful the theory, you should occasionally look at the evidence

**Comment by kyleam** [comment]
> So you use `helm-completing-read-with-cands-in-buffer` as
> `magit-completing-read-function`

No, when I switched from ido to helm not too long ago (Mar 10, according
to my .emacs.d repo), I changed magit-completing-read-function back to
the default value.  When helm is active, it seems to takes over all
completing-read and read-file-name calls (unless something is explicitly
set in helm-completing-read-handlers-alist), eventually calling
helm--completing-read-default and helm--generic-read-file-name,
respectively.

> and that never causes any problems?

I haven't noticed any.  The issue in #1770 seems to be related to using
helm-completing-read-with-cands-in-buffer.

At least with the current helm, I don't think any extra configuration is
needed.
